cranelift/jit/src/memory/arena.rs

```diff
@@ -0,0 +1,218 @@
+        assert!(self.has_space_for(size, align));
+        self.position = align_up(self.position, align);
+        let ptr = unsafe { self.ptr.add(self.position) };
+        self.position += size;
+        ptr
+    }
+
+    fn has_space_for(&self, size: usize, align: usize) -> bool {
+        !self.finalized && align_up(self.position, align) + size <= self.len
+    }
+}
+
+/// `ArenaMemoryProvider` allocates segments from a contiguous memory region
+/// that is reserved up-front.
+///
+/// The arena's memory is initially allocated with PROT_NONE and gradually
+/// updated as the JIT requires more space. This approach allows for stable
+/// addresses throughout the lifetime of the JIT.
+///
+/// Note: Memory will be leaked by default unless
+/// [`JITMemoryProvider::free_memory`] is called to ensure function pointers
+/// remain valid for the remainder of the program's life.
+pub struct ArenaMemoryProvider {
+    alloc: ManuallyDrop<Option<region::Allocation>>,
+    ptr: *mut u8,
+    size: usize,
+    position: usize,
+    segments: Vec<Segment>,
+}
+
+impl ArenaMemoryProvider {
+    /// Create a new memory region with the given size.
+    pub fn new_with_size(reserve_size: usize) -> Result<Self, region::Error> {
+        let size = align_up(reserve_size, region::page::size());
+        let mut alloc = region::alloc(reserve_size, region::Protection::NONE)?;
+        let ptr = alloc.as_mut_ptr();
+
+        Ok(Self {
+            alloc: ManuallyDrop::new(Some(alloc)),
+            segments: Vec::new(),
+            ptr,
+            size,
+            position: 0,
+        })
+    }
+
+    fn allocate(
+        &mut self,
+        size: usize,
+        align: usize,
+        protection: region::Protection,
+    ) -> io::Result<*mut u8> {
+        // Note: Add a fast path without a linear scan over segments here?
+
+        // can we fit this allocation into an existing segment
+        if let Some(segment) = self.segments.iter_mut().find(|seg| {
+            seg.target_prot == protection && !seg.finalized && seg.has_space_for(size, align)
+        }) {
+            return Ok(segment.allocate(size, align));
+        }
+
+        // can we resize the last segment?
+        if let Some(segment) = self.segments.iter_mut().last() {
+            if segment.target_prot == protection && !segment.finalized {
+                let align = align.max(region::page::size());
+                let additional_size = align_up(size, align);
+
+                // if our reserved arena can fit the additional size, extend the
+                // last segment
+                if self.position + additional_size <= self.size {
+                    segment.len += additional_size;
+                    segment.set_rw();
+                    self.position += additional_size;
+                    return Ok(segment.allocate(size, align));
+                }
+            }
+        }
+
+        // allocate new segment for given size and alignment
```

Comment:
Nit: most comments throughout the project follow the convention of capitalizing sentences and ending them with a period.